src/borg/repository.py

```diff
@@ -455,8 +461,22 @@ def complete_xfer(intermediate=True):
                     segments.setdefault(new_segment, 0)
                     segments[new_segment] += 1
                     segments[segment] -= 1
+                elif tag == TAG_PUT and not in_index:
+                    # If this is a PUT shadowed by a later tag, then it will be gone when this segment is deleted after
+                    # this loop. Therefore it is removed from the shadow index.
+                    try:
+                        self.shadow_index[key].remove(segment)
+                    except (KeyError, ValueError):
+                        pass
                 elif tag == TAG_DELETE:
-                    if index_transaction_id is None or segment > index_transaction_id:
+                    # If the shadow index doesn't contain this key, then we can't say if there's a shadowed older tag,
+                    # therefore we do not drop the delete, but write it to a current segment.
+                    shadowed_put_exists = key not in self.shadow_index or any(
+                        # If the key is in the shadow index, then no entry may be older than this tag, otherwise it
+                        # is shadowed.
```

Comment:
if shadowed_put_exists is True, the code below keeps the DELETE (transfers it to new segment).
- same segment (or older segment) PUT older than DELETE: keep the DELETE
- same segment DELETE oder than PUT: keep the DELETE (looks wrong)